packages/cloudvision-grpc-web/src/grpc/index.ts

```diff
@@ -1,80 +1,77 @@
 import { grpc } from '@improbable-eng/grpc-web';
-import { ControlFunctions, GrpcControlMessage, GrpcSource, RpcOptions } from '@types';
-import { defer, finalize, share, Subject } from 'rxjs';
-
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-const DEFAULT_CONTROL_FUNCTIONS: ControlFunctions<any> = {
-  onHeaders(controlMessageSubject, _dataSubject, headers) {
-    controlMessageSubject.next({ metadata: headers });
-  },
-  onMessage(_controlMessageSubject, dataSubject, response) {
-    dataSubject.next(response);
-  },
-  onEnd(controlMessageSubject, dataSubject, code, message) {
-    controlMessageSubject.next({ error: { code, message } });
-    dataSubject.complete();
-    controlMessageSubject.complete();
-  },
-};
+import type { GrpcControlMessage, GrpcEnd, GrpcInvoke, GrpcSource, RpcOptions } from '@types';
+import { map, merge, Observable, Subject } from 'rxjs';
 
 /**
- * Runs [grpc.invoke](https://github.com/improbable-eng/grpc-web/blob/master/client/grpc-web/docs/invoke.md)
- * on the GRPC service and converts the response into two observable streams.
- * One is the `data` stream the other is the `message` stream. The message stream
- * contains GRPC metadata information, as well as any error messages.
- *
- * @param methodDescriptor The GRPC service method definition to be queried.
- * @param options Options to pass to eh GRPC call.
- * @param controlFunctions (optional) Implementation of the onHeaders, onMessage
- * and onEnd GRPC callbacks. This is where publishing to either the `data` or
- * `message` stream is done.
- * @returns An object with the properties `data` and `message`, which are
- * [RXJS Observables](https://rxjs.dev/api/index/class/Observable) that can be subscribed to.
+ * RxJs wrapper around `grpc.invoke` that creates an Observable.
+ * On subscription, the underlying `grpc.invoke` is called.
+ * On unsubscription, the active request is closed.
  */
 export function fromGrpcInvoke<Req extends grpc.ProtobufMessage, Res extends grpc.ProtobufMessage>(
   methodDescriptor: grpc.MethodDefinition<Req, Res>,
   options: RpcOptions<Req, Res>,
-  controlFunctions?: Partial<ControlFunctions<Res>>,
```

Comment:
Discussed offline, there is a risk of breaking compatibility but the current version is stable enough to be used until migration to this. This will be bumped to the next major version